src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs

```diff
@@ -129,117 +128,128 @@ public virtual InputFormatterExceptionPolicy ExceptionPolicy
 
             var suppressInputFormatterBuffering = _options.SuppressInputFormatterBuffering;
 
+            var readStream = request.Body;
             if (!request.Body.CanSeek && !suppressInputFormatterBuffering)
             {
                 // JSON.Net does synchronous reads. In order to avoid blocking on the stream, we asynchronously
                 // read everything into a buffer, and then seek back to the beginning.
-                request.EnableBuffering();
-                Debug.Assert(request.Body.CanSeek);
+                var memoryThreshold = DefaultMemoryThreshold;
+                if (request.ContentLength.HasValue && request.ContentLength.Value > 0 && request.ContentLength.Value < memoryThreshold)
+                {
+                    // If the Content-Length is known and is smaller than the default buffer size, use it.
+                    memoryThreshold = (int)request.ContentLength.Value;
+                }
+
+                readStream = new FileBufferingReadStream(request.Body, memoryThreshold);
 
-                await request.Body.DrainAsync(CancellationToken.None);
-                request.Body.Seek(0L, SeekOrigin.Begin);
+                await readStream.DrainAsync(CancellationToken.None);
+                readStream.Seek(0L, SeekOrigin.Begin);
             }
 
-            using (var streamReader = context.ReaderFactory(request.Body, encoding))
+            var successful = true;
+            Exception exception = null;
+            object model;
+
+            using (var streamReader = context.ReaderFactory(readStream, encoding))
```

Comment:
Just moved the `ErrorHandler` out of the try-catch block since it doesn't need to be there. Besides disposing the `FileBufferingReadStream`, the rest of this is unchanged.